File: openfeature/provider_tests/java-integration-tests/src/test/java/org/gofeatureflag/integrationtests/ProviderTests.java
package org.gofeatureflag.integrationtests;


import dev.openfeature.contrib.providers.gofeatureflag.GoFeatureFlagProvider;
import dev.openfeature.contrib.providers.gofeatureflag.GoFeatureFlagProviderOptions;
import dev.openfeature.contrib.providers.gofeatureflag.exception.InvalidOptions;
import dev.openfeature.sdk.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.function.Consumer;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ProviderTests {
    private static final String relayProxyEndpoint = "http://localhost:1031";
    private static final String relayProxyAuthenticatedEndpoint = "http://localhost:1032";
    private EvaluationContext defaultEvaluationContext;
    private Client goffClient;

    private ImmutableMetadata defaultMetadata = ImmutableMetadata.builder()
            .addString("description", "this is a test")
            .addString("pr_link", "https://github.com/thomaspoignant/go-feature-flag/pull/916")
            .build();
    @BeforeEach
    void init() throws InvalidOptions, ExecutionException, InterruptedException {
        MutableContext userContext = new MutableContext()
                .add("email", "john.doe@gofeatureflag.org")
                .add("firstname", "john")
                .add("lastname", "doe")
                .add("anonymous", false)
                .add("professional", true)
                .add("rate", 3.14)
                .add("age", 30)
                .add("admin", true)
                .add("labels", Arrays.asList(new Value("pro"), new Value("beta")))
                .add("company_info", new MutableStructure().add("name", "my_company").add("size", 120));
        userContext.setTargetingKey("d45e303a-38c2-11ed-a261-0242ac120002");
        defaultEvaluationContext = userContext;

        GoFeatureFlagProviderOptions options = GoFeatureFlagProviderOptions.builder().endpoint(relayProxyEndpoint).build();
        GoFeatureFlagProvider provider = new GoFeatureFlagProvider(options);
        OpenFeatureAPI.getInstance().setProvider(provider);
        OpenFeatureAPI api = OpenFeatureAPI.getInstance();
        goffClient = api.getClient();
        this.waitProviderReady();
    }


    @DisplayName("bool: should resolve a valid boolean flag with TARGETING_MATCH reason")
    @Test
    void shouldResolveAValidBooleanFlagWithTargetingMatchReason() {
        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.TARGETING_MATCH.toString())
                .value(true)
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails(flagKey, false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("bool: should use boolean default value if the flag is disabled")
    @Test
    void shouldUseBooleanDefaultValueIfTheFlagIsDisabled() {
        String flagKey = "disabled_bool";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.DISABLED.toString())
                .value(false)
                .variant("SdkDefault")
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails(flagKey, false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("bool: should error if we expect a boolean and got another type")
    @Test
    void shouldErrorIfWeExpectABooleanAndGotAnotherType() {
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(false)
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("Flag value string_key had unexpected type class java.lang.String, expected class java.lang.Boolean.")
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails("string_key", false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("bool: should error if flag does not exists")
    @Test
    void boolShouldErrorIfFlagDoesNotExists() {
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(false)
                .errorCode(ErrorCode.FLAG_NOT_FOUND)
                .errorMessage("Flag does_not_exists was not found in your configuration")
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails("does_not_exists", false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("string: should resolve a valid string flag with TARGETING_MATCH reason")
    @Test
    void shouldResolveAValidStringFlagWithTargetingMatchReason() {
        String flagKey = "string_key";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.TARGETING_MATCH.toString())
                .flagKey(flagKey)
                .value("CC0000")
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<String> got = goffClient.getStringDetails(flagKey, "default", defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("string: should use string default value if the flag is disabled")
    @Test
    void shouldUseStringDefaultValueIfTheFlagIsDisabled() {
        String flagKey = "disabled_string";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.DISABLED.toString())
                .value("default")
                .variant("SdkDefault")
                .build();
        FlagEvaluationDetails<String> got = goffClient.getStringDetails(flagKey, "default", defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("string: should error if we expect a string and got another type")
    @Test
    void shouldErrorIfWeExpectAStringAndGotAnotherType() {
        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value("default")
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("Flag value bool_targeting_match had unexpected type class java.lang.Boolean, expected class java.lang.String.")
                .build();
        FlagEvaluationDetails<String> got = goffClient.getStringDetails(flagKey, "default", defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("string: should error if flag does not exists")
    @Test
    void stringShouldErrorIfFlagDoesNotExists() {
        String flagKey = "does_not_exists";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value("default")
                .errorCode(ErrorCode.FLAG_NOT_FOUND)
                .errorMessage("Flag does_not_exists was not found in your configuration")
                .build();
        FlagEvaluationDetails<String> got = goffClient.getStringDetails(flagKey, "default", defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("double: should resolve a valid double flag with TARGETING_MATCH reason")
    @Test
    void shouldResolveAValidDoubleFlagWithTargetingMatchReason() {
        String flagKey = "double_key";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.TARGETING_MATCH.toString())
                .flagKey(flagKey)
                .value(100.25)
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<Double> got = goffClient.getDoubleDetails(flagKey, 123.45, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("double: should use double default value if the flag is disabled")
    @Test
    void shouldUseDoubleDefaultValueIfTheFlagIsDisabled() {
        String flagKey = "disabled_float";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.DISABLED.toString())
                .value(123.45)
                .variant("SdkDefault")
                .build();
        FlagEvaluationDetails<Double> got = goffClient.getDoubleDetails(flagKey, 123.45, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("double: should error if we expect a double and got another type")
    @Test
    void shouldErrorIfWeExpectADoubleAndGotAnotherType() {
        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(123.45)
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("Flag value bool_targeting_match had unexpected type class java.lang.Boolean, expected class java.lang.Double.")
                .build();
        FlagEvaluationDetails<Double> got = goffClient.getDoubleDetails(flagKey, 123.45, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("double: should error if flag does not exists")
    @Test
    void doubleShouldErrorIfFlagDoesNotExists() {
        String flagKey = "does_not_exists";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(123.45)
                .errorCode(ErrorCode.FLAG_NOT_FOUND)
                .errorMessage("Flag does_not_exists was not found in your configuration")
                .build();
        FlagEvaluationDetails<Double> got = goffClient.getDoubleDetails(flagKey, 123.45, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("int: should resolve a valid integer flag with TARGETING_MATCH reason")
    @Test
    void shouldResolveAValidIntFlagWithTargetingMatchReason() {
        String flagKey = "integer_key";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.TARGETING_MATCH.toString())
                .flagKey(flagKey)
                .value(100)
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<Integer> got = goffClient.getIntegerDetails(flagKey, 123, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("int: should use integer default value if the flag is disabled")
    @Test
    void shouldUseIntDefaultValueIfTheFlagIsDisabled() {
        String flagKey = "disabled_int";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.DISABLED.toString())
                .value(123)
                .variant("SdkDefault")
                .build();
        FlagEvaluationDetails<Integer> got = goffClient.getIntegerDetails(flagKey, 123, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("int: should error if we expect a integer and got another type")
    @Test
    void shouldErrorIfWeExpectAIntAndGotAnotherType() {
        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(123)
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("Flag value bool_targeting_match had unexpected type class java.lang.Boolean, expected class java.lang.Integer.")
                .build();
        FlagEvaluationDetails<Integer> got = goffClient.getIntegerDetails(flagKey, 123, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("int: should error if flag does not exists")
    @Test
    void intShouldErrorIfFlagDoesNotExists() {
        String flagKey = "does_not_exists";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(123)
                .errorCode(ErrorCode.FLAG_NOT_FOUND)
                .errorMessage("Flag does_not_exists was not found in your configuration")
                .build();
        FlagEvaluationDetails<Integer> got = goffClient.getIntegerDetails(flagKey, 123, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("object: should resolve a valid object flag with TARGETING_MATCH reason")
    @Test
    void shouldResolveAValidObjectFlagWithTargetingMatchReason() {
        String flagKey = "object_key";
        Structure expectedValue = new MutableStructure().add("test4", 1).add("test2", false).add("test3", 123.3).add("test", "test1");
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.TARGETING_MATCH.toString())
                .flagKey(flagKey)
                .value(new Value(expectedValue))
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<Value> got = goffClient.getObjectDetails(flagKey, new Value(123), defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("object: should use object default value if the flag is disabled")
    @Test
    void shouldUseObjectDefaultValueIfTheFlagIsDisabled() {
        String flagKey = "disabled_int";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.DISABLED.toString())
                .value(new Value(123))
                .variant("SdkDefault")
                .build();
        FlagEvaluationDetails<Value> got = goffClient.getObjectDetails(flagKey, new Value(123), defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("object: should error if flag does not exists")
    @Test
    void objectShouldErrorIfFlagDoesNotExists() {
        String flagKey = "does_not_exists";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .errorCode(ErrorCode.FLAG_NOT_FOUND)
                .errorMessage("Flag does_not_exists was not found in your configuration")
                .value(new Value(123))
                .build();
        FlagEvaluationDetails<Value> got = goffClient.getObjectDetails(flagKey, new Value(123), defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("authenticated relay proxy: valid")
    @Test
    void authenticatedRelayProxyValid() throws InvalidOptions, ExecutionException, InterruptedException {
        GoFeatureFlagProviderOptions options = GoFeatureFlagProviderOptions.builder()
                .apiKey("authorized_token").endpoint(relayProxyAuthenticatedEndpoint).build();
        GoFeatureFlagProvider provider = new GoFeatureFlagProvider(options);
        OpenFeatureAPI.getInstance().setProvider(provider);
        OpenFeatureAPI api = OpenFeatureAPI.getInstance();
        goffClient = api.getClient();
        this.waitProviderReady();

        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .flagKey(flagKey)
                .reason(Reason.TARGETING_MATCH.toString())
                .value(true)
                .variant("True")
                .flagMetadata(defaultMetadata)
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails(flagKey, false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    @DisplayName("authenticated relay proxy: empty api key")
    @Test
    void authenticatedRelayProxyEmptyToken() throws InvalidOptions, ExecutionException, InterruptedException {
        GoFeatureFlagProviderOptions options = GoFeatureFlagProviderOptions.builder()
                .apiKey("").endpoint(relayProxyAuthenticatedEndpoint).build();
        GoFeatureFlagProvider provider = new GoFeatureFlagProvider(options);
        OpenFeatureAPI.getInstance().setProvider(provider);
        OpenFeatureAPI api = OpenFeatureAPI.getInstance();
        goffClient = api.getClient();
        this.waitProviderReady();

        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(false)
                .errorCode(ErrorCode.GENERAL)
                .errorMessage("impossible to contact GO Feature Flag relay proxy instance")
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails(flagKey, false, defaultEvaluationContext);
        assertEquals(expected, got);

    }

    @DisplayName("authenticated relay proxy: invalid api key")
    @Test
    void authenticatedRelayProxyInvalidToken() throws InvalidOptions, ExecutionException, InterruptedException {
        GoFeatureFlagProviderOptions options = GoFeatureFlagProviderOptions.builder()
                .apiKey("invalid-api-key").endpoint(relayProxyAuthenticatedEndpoint).build();
        GoFeatureFlagProvider provider = new GoFeatureFlagProvider(options);
        OpenFeatureAPI.getInstance().setProvider(provider);
        OpenFeatureAPI api = OpenFeatureAPI.getInstance();
        goffClient = api.getClient();
        this.waitProviderReady();

        String flagKey = "bool_targeting_match";
        FlagEvaluationDetails expected = FlagEvaluationDetails.builder()
                .reason(Reason.ERROR.toString())
                .value(false)
                .errorCode(ErrorCode.GENERAL)
                .errorMessage("invalid token used to contact GO Feature Flag relay proxy instance")
                .build();
        FlagEvaluationDetails<Boolean> got = goffClient.getBooleanDetails(flagKey, false, defaultEvaluationContext);
        assertEquals(expected, got);
    }

    private void waitProviderReady() throws ExecutionException, InterruptedException {
        CompletableFuture<EventDetails> completableFuture = new CompletableFuture<>();
        OpenFeatureAPI.getInstance().onProviderReady(new Consumer<EventDetails>() {
            @Override
            public void accept(EventDetails eventDetails) {
                completableFuture.complete(eventDetails);
            }
        });
        completableFuture.get();
    }
}

File: openfeature/provider_tests/dotnet-integration-tests/ProviderTests/ProviderTests.cs
using System.Runtime.InteropServices.JavaScript;
using OpenFeature.Model;
using OpenFeature;
using OpenFeature.Constant;
using OpenFeature.Contrib.Providers.GOFeatureFlag;
using FluentAssertions;

namespace ProviderTests;


public class ProviderTests
{
    private GoFeatureFlagProvider goFeatureFlagProvider;
    private FeatureClient client;
    private EvaluationContext defaultEvaluationContext;
    [SetUp]
    public void Setup()
    {
        goFeatureFlagProvider = new GoFeatureFlagProvider(new GoFeatureFlagProviderOptions
        {
            Endpoint = "http://localhost:1031/",
            Timeout = new TimeSpan(1000 * TimeSpan.TicksPerMillisecond)
        });
        Api.Instance.SetProvider(goFeatureFlagProvider);
        client = Api.Instance.GetClient("my-app");
        
        defaultEvaluationContext = EvaluationContext.Builder()
            .Set("targetingKey", "d45e303a-38c2-11ed-a261-0242ac120002") // user unique identifier (mandatory)
            .Set("email", "john.doe@gofeatureflag.org")
            .Set("firstname", "john")
            .Set("lastname", "doe")
            .Set("anonymous", false)
            .Set("professional", true)
            .Set("rate", 3.14)
            .Set("age", 30)
            .Set("admin", true)
            .Set("labels", new Value(new List<Value> { new("pro"), new("beta") }))
            .Set("company_info", Structure.Builder().Set("name","my_company").Set("size",120).Build())
            .Build();
    }

    [Test]
    public async Task ShouldResolveAValidBooleanFlagWithTargetingMatchReason()
    {
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            true, 
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null);
        var got = await client.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);

    }
    [Test]
    public async Task ShouldUseBooleanDefaultValueIfTheFlagIsDisabled()
    {
        var flagKey = "disabled_bool";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            false, 
            ErrorType.None, 
            Reason.Disabled,
            null,
            null); 
        var got = await client.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfWeExpectABooleanAndGotAnotherType()
    {
        var flagKey = "string_key";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            false, 
            ErrorType.TypeMismatch, 
            Reason.Error,
            "",
            "flag value string_key had unexpected type"); 
        var got = await client.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfFlagDoesNotExistsBool()
    {
        var flagKey = "does_not_exists";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            false, 
            ErrorType.FlagNotFound, 
            Reason.Error,
            "",
            $"flag {flagKey} was not found in your configuration"); 
        var got = await client.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldResolveAValidStringFlagWithTargetingMatchReason()
    {
        var flagKey = "string_key";
        var want = new FlagEvaluationDetails<string>(
            flagKey, 
            "CC0000", 
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null); 
        var got = await client.GetStringDetails(flagKey, "default", defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldUseStringDefaultValueIfTheFlagIsDisabled()
    {
        var flagKey = "disabled_string";
        var want = new FlagEvaluationDetails<string>(
            flagKey, 
            "default", 
            ErrorType.None, 
            Reason.Disabled,
            null,
            null); 
        var got = await client.GetStringDetails(flagKey, "default", defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfWeExpectAStringAndGotAnotherType()
    {
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<string>(
            flagKey, 
            "default", 
            ErrorType.TypeMismatch, 
            Reason.Error,
            "",
            "flag value bool_targeting_match had unexpected type"); 
        var got = await client.GetStringDetails(flagKey, "default", defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfFlagDoesNotExistsString()
    {
        var flagKey = "does_not_exists";
        var want = new FlagEvaluationDetails<string>(
            flagKey, 
            "default", 
            ErrorType.FlagNotFound, 
            Reason.Error,
            "",
            $"flag {flagKey} was not found in your configuration"); 
        var got = await client.GetStringDetails(flagKey, "default", defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldResolveAValidDoubleFlagWithTargetingMatchReason()
    {
        var flagKey = "double_key";
        var want = new FlagEvaluationDetails<double>(
            flagKey, 
            100.25,
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null); 
        var got = await client.GetDoubleDetails(flagKey,  123.45, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldUseDoubleDefaultValueIfTheFlagIsDisabled()
    {
        var flagKey = "disabled_float";
        var want = new FlagEvaluationDetails<double>(
            flagKey, 
            123.45,
            ErrorType.None, 
            Reason.Disabled,
            null,
            null); 
        var got = await client.GetDoubleDetails(flagKey,  123.45, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfWeExpectADoubleAndGotAnotherType()
    {
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<double>(
            flagKey, 
            123.45,
            ErrorType.TypeMismatch, 
            Reason.Error,
            "",
            "flag value bool_targeting_match had unexpected type"); 
        var got = await client.GetDoubleDetails(flagKey,  123.45, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfFlagDoesNotExistsDouble()
    {
        var flagKey = "does_not_exists";
        var want = new FlagEvaluationDetails<double>(
            flagKey, 
            123.45,
            ErrorType.FlagNotFound, 
            Reason.Error,
            "",
            $"flag {flagKey} was not found in your configuration"); 
        var got = await client.GetDoubleDetails(flagKey, 123.45, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldResolveAValidIntFlagWithTargetingMatchReason()
    {
        var flagKey = "integer_key";
        var want = new FlagEvaluationDetails<int>(
            flagKey, 
            100,
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null); 
        var got = await client.GetIntegerDetails(flagKey,  123, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldUseIntDefaultValueIfTheFlagIsDisabled()
    {
        var flagKey = "disabled_int";
        var want = new FlagEvaluationDetails<int>(
            flagKey, 
            123,
            ErrorType.None, 
            Reason.Disabled,
            null,
            null); 
        var got = await client.GetIntegerDetails(flagKey,  123, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfWeExpectAIntAndGotAnotherType()
    {
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<int>(
            flagKey, 
            123,
            ErrorType.TypeMismatch, 
            Reason.Error,
            "",
            "flag value bool_targeting_match had unexpected type"); 
        var got = await client.GetIntegerDetails(flagKey,  123, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfFlagDoesNotExistsInt()
    {
        var flagKey = "does_not_exists";
        var want = new FlagEvaluationDetails<int>(
            flagKey, 
            123,
            ErrorType.FlagNotFound, 
            Reason.Error,
            "",
            $"flag {flagKey} was not found in your configuration"); 
        var got = await client.GetIntegerDetails(flagKey,  123, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldResolveAValidObjectFlagWithTargetingMatchReason()
    {
        Structure expectedStructure = new Structure(
        new Dictionary<string, Value>()
            {
                { "test4", new Value(1) },
                { "test2", new Value(false) },
                {"test3", new Value(123.3)},
                {"test", new Value("test1")},
            }
        );
        var flagKey = "object_key";
        var want = new FlagEvaluationDetails<Value>(
            flagKey, 
            new Value(expectedStructure),
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null); 
        var got = await client.GetObjectDetails(flagKey,  new Value(123), defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldUseObjectDefaultValueIfTheFlagIsDisabled()
    {
        var flagKey = "disabled_int";
        var want = new FlagEvaluationDetails<Value>(
            flagKey, 
            new Value(123),
            ErrorType.None, 
            Reason.Disabled,
            null,
            null); 
        var got = await client.GetObjectDetails(flagKey,  new Value(123), defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task ShouldErrorIfFlagDoesNotExistsObject()
    {
        var flagKey = "does_not_exists";
        var want = new FlagEvaluationDetails<Value>(
            flagKey, 
            new Value(123),
            ErrorType.FlagNotFound, 
            Reason.Error,
            "",
            $"flag {flagKey} was not found in your configuration"); 
        var got = await client.GetObjectDetails(flagKey,  new Value(123), defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task Should_resolve_a_valid_flag_with_an_apiKey()
    {
        GoFeatureFlagProvider authenticatedProvider = new GoFeatureFlagProvider(new GoFeatureFlagProviderOptions
        {
            Endpoint = "http://localhost:1032/",
            Timeout = new TimeSpan(1000 * TimeSpan.TicksPerMillisecond),
            ApiKey = "authorized_token"
        });
        Api.Instance.SetProvider(authenticatedProvider);
        FeatureClient authenticatedClient = Api.Instance.GetClient("my-app");
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            true, 
            ErrorType.None, 
            Reason.TargetingMatch,
            "True",
            null);
        var got = await authenticatedClient.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    
    [Test]
    public async Task Should_resolve_a_default_value_with_an_invalid_apiKey()
    {
        GoFeatureFlagProvider authenticatedProvider = new GoFeatureFlagProvider(new GoFeatureFlagProviderOptions
        {
            Endpoint = "http://localhost:1032/",
            Timeout = new TimeSpan(1000 * TimeSpan.TicksPerMillisecond),
            ApiKey = "invalid_api_key"
        });
        Api.Instance.SetProvider(authenticatedProvider);
        FeatureClient authenticatedClient = Api.Instance.GetClient("my-app");
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            false, 
            ErrorType.General, 
            Reason.Error,
            "True",
            "invalid api key, impossible to authenticate the provider");
        var got = await authenticatedClient.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
    [Test]
    public async Task Should_resolve_a_default_value_with_an_empty_apiKey()
    {
        GoFeatureFlagProvider authenticatedProvider = new GoFeatureFlagProvider(new GoFeatureFlagProviderOptions
        {
            Endpoint = "http://localhost:1032/",
            Timeout = new TimeSpan(1000 * TimeSpan.TicksPerMillisecond),
            ApiKey = ""
        });
        Api.Instance.SetProvider(authenticatedProvider);
        FeatureClient authenticatedClient = Api.Instance.GetClient("my-app");
        var flagKey = "bool_targeting_match";
        var want = new FlagEvaluationDetails<bool>(
            flagKey, 
            false, 
            ErrorType.General, 
            Reason.Error,
            "True",
            "invalid api key, impossible to authenticate the provider");
        var got = await authenticatedClient.GetBooleanDetails(flagKey, false, defaultEvaluationContext);
        want.Should().BeEquivalentTo(got);
    }
}